compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ClassCodegen.kt

```diff
@@ -313,79 +310,28 @@ open class ClassCodegen protected constructor(
 
 }
 
-fun ClassDescriptor.calculateClassFlags(): Int {
-    var flags = 0
-    flags = flags or if (JvmCodegenUtil.isJvmInterface(this)) Opcodes.ACC_INTERFACE else Opcodes.ACC_SUPER
-    flags = flags or calcModalityFlag()
-    flags = flags or AsmUtil.getVisibilityAccessFlagForClass(this)
-    flags = flags or if (kind == ClassKind.ENUM_CLASS) Opcodes.ACC_ENUM else 0
-    flags = flags or if (kind == ClassKind.ANNOTATION_CLASS) Opcodes.ACC_ANNOTATION else 0
-    return flags
-}
-
-fun MemberDescriptor.calculateCommonFlags(): Int {
-    var flags = 0
-    if (Visibilities.isPrivate(visibility)) {
-        flags = flags.or(Opcodes.ACC_PRIVATE)
-    } else if (visibility == Visibilities.PUBLIC || visibility == Visibilities.INTERNAL) {
-        flags = flags.or(Opcodes.ACC_PUBLIC)
-    } else if (visibility == Visibilities.PROTECTED) {
-        flags = flags.or(Opcodes.ACC_PROTECTED)
-    } else if (visibility == JavaVisibilities.PACKAGE_VISIBILITY) {
-        // default visibility
-    } else {
-        throw RuntimeException("Unsupported visibility $visibility for descriptor $this")
+private val IrClass.flags: Int
+    get() = origin.flags or AsmUtil.getVisibilityAccessFlagForClass(descriptor) or when {
+        isAnnotationClass -> Opcodes.ACC_ANNOTATION or Opcodes.ACC_INTERFACE or Opcodes.ACC_ABSTRACT
+        isEnumClass -> Opcodes.ACC_ENUM
```

Comment:
If there are no classes for entries enum class should be FINAL

Reply:
Huh, the code didn't do that previously. `or`-ing ACC_ENUM with modality flags actually fixes a test.

Reply:
@pyos I can't find ACC_SUPER anymore(
Actually tests for flags are located under 'writeFlags' folder (see AbstractWriteFlagsTest) and Ir analogue is not implemented

Reply:
Ouch, I somehow lost ACC_SUPER, fixed.

Reply:
Could you also add test for 'writeFlags', it should be simple by AbstractIrBlackBoxCodegenTest analogue

Reply:
@pyos KotlinTestUtils. AUTOMATICALLY_MUTE_FAILED_TESTS should significally help. 
Please send new tests as separate pr

Reply:
https://github.com/JetBrains/kotlin/pull/2251

The enum-related change fixes 6 tests, but I haven't found any tests with ACC_SYNTHETIC that aren't failing for other reasons, such as different class names.